R/ZIPLNfit-class.R

```diff
@@ -0,0 +1,831 @@
+          message     = stop_reason,
+          objective   = criterion,
+          convergence = convergence)
+      )
+    },
+
+    #' @description Predict position, scores or observations of new data.
+    #' @param newdata A data frame in which to look for variables with which to predict. If omitted, the fitted values are used.
+    #' @param responses Optional data frame containing the count of the observed variables (matching the names of the provided as data in the PLN function), assuming the interest in in testing the model.
+    #' @param type Scale used for the prediction. Either `link` (default, predicted positions in the latent space) or `response` (predicted counts).
+    #' @param level Optional integer value the level to be used in obtaining the predictions. Level zero corresponds to the population predictions (default if `responses` is not provided) while level one (default) corresponds to predictions after evaluating the variational parameters for the new data.
+    #' @param envir Environment in which the prediction is evaluated
+    #'
+    #' @details
+    #' Note that `level = 1` can only be used if responses are provided,
+    #' as the variational parameters can't be estimated otherwise. In the absence of responses, `level` is ignored and the fitted values are returned
+    #' @return A matrix with predictions scores or counts.
+    predict = function(newdata, responses = NULL, type = c("link", "response"), level = 1, envir = parent.frame()) {
+
+      ## Ignore everything if newdata is not provided
+      if (missing(newdata)) {
+        return(self$fitted)
+      }
+
+      n_new <- nrow(newdata)
+      ## Set level to 0 (to bypass VE step) if responses are not provided
+      if (is.null(responses)) {
+        level <- 0
+      }
+
+      terms <- .extract_terms_zi(as.formula(private$formula))
+
+      ## Extract the model matrices from the new data set with initial formula
+      X <- model.matrix(terms$PLN[-2], newdata, xlev = attr(private$formula, "xlevels")$PLN)
+      if (!is.null(terms$ZI)) X0 <- model.matrix(terms$ZI, newdata, xlev = attr(private$formula, "xlevels")$ZI) else X0 <- matrix(NA,0,0)
+
+      O <- model.offset(model.frame(terms$PLN[-2], newdata))
+      if (is.null(O)) O <- matrix(0, n_new, self$p)
+
+      ## Optimize M and S if responses are provided,
+      if (level == 1) {
+        VE <- self$optimize_vestep(
+          covariates = list(PLN = X, ZI = X0),
+          offsets    = O,
+          responses  = as.matrix(responses),
+          weights    = rep(1, n_new),
+          B          = private$B,
+          B0         = private$B0,
+          Omega      = private$Omega
+        )
+        R <- VE$R
+        M <- VE$M
+        S <- VE$S
+      } else {
+        # otherwise set R to Pi, M to 0 and S to diag(Sigma)
+        R <- private$Pi[1:nrow(newdata), ]
+        M <- X %*% private$B
+        S <- matrix(diag(private$Sigma), nrow = n_new, ncol = self$p, byrow = TRUE)
```

Comment:
`S` is the matrix of variational standard devations (and `S^2` of variances) so we should use `sqrt(diag(private$Sigma))` instead of `sqrt(diag(private$Sigma))`

Reply:
you're right, that is a mistake. It's not likely to change the world, though...